pkg/mempool/simplemempool/internal/parts/formbatches/formbatches.go

```diff
@@ -26,6 +26,7 @@ func IncludeBatchCreation(
 		State:    commonState,
 		NewTxIDs: nil,
 	}
+	txIDCount := make(map[string]int, 0)
```

Comment:
Yes that is the reason. I am aware that with the current implementation the app logic is responsible for filtering collisions and just keeping one copy, but here the purpose is to prevent the state from blowing app by keeping in memory all transactions (i.e. removing after decision). The fix is clear. However, if I remove a transaction after the first decision and that transaction is proposed again (by a malicious proposer), then, after deciding on this proposal, line `txIDs = append(txIDs, txID)` will throw a runtime error, because `tx := state.TxByID[string(txID)]` will return nil (and `append(slice, nil)` throws a runtime error.

Reply:
I think there are 2 concepts being mixed here: (1) proposing transactions and (2) adding transactions to the mempool.

Unless I missed something (possible), the mempool is a purely local data structure that serves as a buffer for transactions waiting to be proposed by _this node_, regardless of other malicious nodes are proposing. Or does the mempool get involved in any way during other nodes' proposals? It is possible, though, that some client (not even necessarily a malicious one) can send the same transaction to the same node more than once.

But this should be addressed differently in my opinion. It would be both more natural and way simpler to just keep at most one copy of a transaction in the mempool.
1. Remove the counter
2. When adding transactions, ignore those that are already inside according to the index (note that this is easy, as everything is executed sequentially)

With this approach, is there still a problem that the counter would solve?

Reply:
Actually yes you are absolutely right, I thought the mempool was storing transactions received from batches by the availability module, which in this case is of course the responsibility of the fakebatchdb (and obviously the fakebatchdb is already known to blow up the state by keeping the database in plain maps in memory). I just updated the code to remove the counter on the txID (and assume as elsewhere that the app logic prevents collisions).

Reply:
What does need to be fixed, though, is checking for the presence of the transaction when adding it, and only add it if it is not yet in the `TxByID` index (otherwise ignore it).

Reply:
done